src/sentry/ratelimits/sliding_windows.py

```diff
@@ -0,0 +1,367 @@
+"""
+The ratelimiter used by the metrics string indexer to rate-limit DB writes.
+
+As opposed to the API rate limiter, a limit in the "sliding window" rate
+limiter such as "10 requests / minute" does not reset to 0 every minute.
+Instead each window can be configured with a "granularity" setting such that
+the window gradually resets in steps of `granularity` seconds.
+
+Additionally this rate-limiter is not coupled to per-project/organization
+scopes, and can apply multiple sliding windows at once. On the flipside it is
+not strongly consistent and depending on usage it is very easy to over-spend
+quota, as checking quota and spending quota are two separate steps.
+
+Example
+=======
+
+We want to enforce the number of requests per organization both per-hour
+(100/hour) and per-minute (10/minute). On every request, our API endpoint
+calls:
+
+    check_and_use_quotas(RequestedQuota(
+        prefix=f"org-id:{org_id}"
+        quotas=[
+            Quota(
+                window_seconds=3600,
+                limit=100,
+
+                # can be arbitrary depending on how "sliding" the sliding
+                # window should be. This one configures per-second granularity
+                # to make the example simpler
+                granularity=1,
+            ),
+            Quota(
+                window_seconds=60,
+                limit=10,
+                granularity=10,
+            )
+        ]
+    ))
+
+For a request happening at time `900` for `org_id=123`, the redis backend
+checks the following keys::
+
+    sliding-window-rate-limit:123:60:900
+    sliding-window-rate-limit:123:60:899
+    sliding-window-rate-limit:123:60:898
+    ...
+
+    sliding-window-rate-limit:123:3600:90
+    sliding-window-rate-limit:123:3600:89
+    sliding-window-rate-limit:123:3600:88
+    ...
+
+...none of which exist, so the values are assumed 0 and the request goes
+through. It then sets the following keys:
+
+    sliding-window-rate-limit:123:3600:90 += 1
+    sliding-window-rate-limit:123:60:900 += 1
+
+After one minute, another request for the same org happens at time `902`, and
+the keys change as follows:
+
+    sliding-window-rate-limit:123:60:900 = 1
+    sliding-window-rate-limit:123:60:902 = 1
+    sliding-window-rate-limit:123:3600:90 = 2
```

Comment:
I think it's worth explaining the non-empty case as that is more likely to be the usual state of things. Namely that:

* the used quota for the window is calculated by summing the granules
* that the smallest remaining quota will be the one that is granted

Reply:
ok, i added a little bit more info